### ISSUE TITLE
sqlilte-gui roadmap

### ISSUE THREAD

**Issue by SilvioGrosso** [little-brother/sqlite-gui#84]
Hello @little-brother!

Thanks a lot for releasing sqlite-gui.
Just tried it on Windows 10 and it works great.
No crash whatsoever so far :-)
I have always worked with DB browser for SQlite (https://sqlitebrowser.org/) so far which is incredibly powerful

Do you plan to add some new features in the future?
More precisely, is there somewhere a roadmap as regarding your sqlite-gui?

For instance, would be possible to add in some future version the:
**extension-functions.c** (50.96 KB) contributed by Liam Healy on 2010-02-06 15:45:07 (https://www.sqlite.org/contrib)
I have noticed it is not included right now by default (just tried its reverse option, for reverting strings, and it is not recognized by sqlite-gui)

Thanks again!

**Comment by little-brother** [comment]
Hi.

`extension-functions.c `-extension has been removed since 1.6.0 because some (math) of functions  are implemented in SQLite 3.36.0 by default. You can copy dll from DB4S to `extensions`-folder if there are missing function.  

> Do you plan to add some new features in the future?
> More precisely, is there somewhere a roadmap
I have ToDo-list. There are many points but this is not a real roadmap :) 
You can suggest new ideas and maybe they will be implemented.
The best way to get a new features is to report a bug, because the new release contains all current bug fixes and new features.

**Comment by SilvioGrosso** [comment]
Hello @little-brother 

First off, thanks a lot indeed for your reply :-)
I have just added math.dll (taken by Db browser...) to Sqlite-gui and not I can use the reverse function together with all the others available

As regards new features I leave to you their choice. As of today, sqlite-gui is already quite **impressive**!
I have noticed it is unable to show images (jpg, png) in the GUIs.
If I open a database with **blob**, as images (e.g. one created with Db Browser), they are not "visible" in the GUIs. 
Whenever I close the  window, by pressing X (left upper icon) within the table with the blob images, the application crashes. 
In the CMD prompt window there are no errors reported related to this crash.
Tested on Windows 64 (1.6.5 version)

**Comment by little-brother** [comment]
Could you upload this database here?

> I have noticed I am unable to show images

Maybe there is unsupported format. Images should be visible in preview

![изображение](https://user-images.githubusercontent.com/20204037/146683644-12598ce2-e1b6-499c-a823-3cef95dcc417.png)

**Comment by SilvioGrosso** [comment]
Hello @little-brother 

Ok. It works now:
<img width="960" alt="SQLITE-GUI" src="https://user-images.githubusercontent.com/33007383/146687013-baace93f-47ff-4d13-80f1-2d393fec62c9.png">

Just out of curiosity, I have also tried to open images from my disk and it works:
select * **from fsdir** ('D:\prova.jpg');

As regards the crash, I have tried again to reproduce but now Sqlite-gui works fine :)
I suppose it was Db browser for sqltie _fault_ since, with this software, you can drag and drop an image (jpg, png) into the table (the blob cell) and maybe Sqlite-gui was unable to read this blob (its code) afterwards...

Please, keep up your work on Sqlite-gui. 
It is indeed useful and extremly ypowerful! :)

**Comment by little-brother** [comment]
>  it was Db browser for sqltie fault since

Nope. There is a bug in sqlite-gui definitely. It will be great if you can reproduce this error and provide more details. Perhaps, I need to add some checks to code to prevent an application crush on opening broken images.

**Comment by SilvioGrosso** [comment]
Ok.

It suffices to close the table and, sooner or later, the crush occurs.
The image (blob) has been imported previously (drag and drop) with db browser where it is showed correctly (no crash)

Here is the  video on Windows 10 where I show all passages (it crashes the 3-4° time I close the table...):

https://user-images.githubusercontent.com/33007383/146689299-a80645ef-7bfb-471f-8d4c-a10d26faadc4.mp4

Here is the database (stored on One Drive; NO need to have an account to download it...):
https://1drv.ms/u/s!AjEV_9wBPHINgoV5vHwFpLr0NMselg?e=pjeAMk

**Comment by SilvioGrosso** [comment]
Hello @little-brother 

I have tried to visualize the same image with Sqltiestudio (https://sqlitestudio.pl/)
It doesn't crash but it is unable to show the jpeg image.

<img width="960" alt="sqlitestudio" src="https://user-images.githubusercontent.com/33007383/146689775-7489fed1-d73f-4b33-88c3-573e0c581686.png">

Therefore the only software which works 100% is db browser so far :-)

BTW, in the long past, I worked with SqliteStudio. 
Later on I have complelty switched to Db Browser since, in my view, it is more powerful. Unfortunately, its development has nearly stopped right now. For instance, currently, Db browser does not support strict tables (SQlite 3.37.0  new feature)

**Comment by little-brother** [comment]
Thanks a lot! Now I can try to fix this issue.

> DB Browser development has nearly stopped right now. For instance, currently, it does not support strict tables (SQlite 3.37.0 new feature)

I think DB4S have a lack of ideas (not enthusiasm, I hope). At least sqlite-gui has :) There is a big problem to find a feature that will be used the majority of users. An excess of functionality is as bad as a lack of it.

Both DB4S and SQLiteStrudio are great applications, but imho they have overcomplicated interface.

**Comment by little-brother** [comment]
Heh, DB4S use incorrect data type to store BLOB. Somehow it use `TEXT` instead of `BLOB` (execute `select typeof(img) fro images` --> `TEXT`). That is why nor `sqlite-gui`, nor `SQLite Studio` don't show image.
I have tried to repeat this behaviour on a new database without luck (DB4S stored images as `BLOB`).

**Comment by SilvioGrosso** [comment]
Hello!

Did you reproduce the crash?
On my system (Windows 10) it crashes regulararly after closing a few times, in the row, the table with the blob

BTW, I have tried to import a CSV (created by LibreOffice Calc) file to no avail.
Sqlite-gui always throws an error (with no hint to understand the problem...) and it is unable to import these data.
With the same CSV Db Browser and SqliteStudio have no problem whatsoever. They import all rows into the table

If you are interested I can open a new issue :-)
 

    Il domenica 19 dicembre 2021, 23:52:15 CET, little-brother ***@***.***> ha scritto:  
 
 


Heh, DB4S use incorrect data type to store BLOB. Somehow it use TEXT instead of BLOB (execute select typeof(img) fro images --> TEXT). That is why nor sqlite-gui, nor SQLite Studio don't show image.
I have tried to repeat this behaviour on a new database without luck (DB4S stored images as BLOB).

—
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.

**Comment by little-brother** [comment]
> If you are interested I can open a new issue
Yes. And please attach the csv-file if it possible.

> Did you reproduce the crash?
Not yet. Perhaps I should load Win10 (I use Win7).
Which version (x32 or x64) is crushing? Or both?

**Comment by SilvioGrosso** [comment]
Hello

Yesterday, at home, sqlite-gui crashed on Windows 10 (home edition) with the X64 version
Today, I have just tried on a very different computer, namely at work, with Windows 10 Pro (X64 version of Sqlite-gui).Here (W10 pro)  it crashes as well. It only takes much MORE efforts :-)

In short, I had to close and reopen the table several times before the application crashed without warnings :-)
I have used the same databse (IMAGES.db).
For instance, when it crashed again, I had to close the table 11 times to unleash it.

IMHO, it doesn't merit much investigation since it might be a "corner case" difficult to debug :-)
 

    Il lunedì 20 dicembre 2021, 16:11:28 CET, little-brother ***@***.***> ha scritto:  
 
 



If you are interested I can open a new issue
Yes. And please attach the csv-file if it possible.



Did you reproduce the crash?
Not yet. Perhaps I should load Win10 (I use Win7).
Which version (x32 or x64) is crushing? Or both?


—
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.

**Comment by little-brother** [comment]
It looks like a buffer overflow. The blob data stored as `TEXT` and that is why it's possible to edit it (which is undoubtedly pointless). When you switc to another field the application tried to update BLOB value with incorrect data (the edit-control displays binary data not properly) and doesn't allocate enough memory for it.

**Comment by SilvioGrosso** [comment]
Yeah. It might be indeed so :-)

At work, I have got a much more powerfull computer running Windows 10 - Pro.
16 gb of Ram (instead of 8 gb, at home).
Also the Intel CPU (i5 8500) is much better compared to yesterday's CPU Intel hardware :-)
 

    Il lunedì 20 dicembre 2021, 16:33:48 CET, little-brother ***@***.***> ha scritto:  
 
 


It looks like a buffer overflow. The blob data stored as TEXT and that is why it's possible to edit it (which is undoubtedly pointless). When you switc to another field the application tried to update BLOB value with incorrect data (the edit-control displays binary data not properly) and doesn't allocate enough memory for it.

—
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.
